### PR TITLE
`GET` route for fetching user device info for qr-code mobile auth

### LINKED ISSUE #1011: Feature Request : Retrieving QR-Code-Auth Document
**Description:**
Implement a GET API call to retrieve the QR-code-auth document to retrieveEndpoint: GET /auth/qr-code-auth

**API Request**
GET

**Endpoint**
/auth/qr-code-auth

**Query Parameter**:
`device_id: [Specify the device ID you used for the request]
`
**Expected Response**

```
{
  "message": "Authentication document retrieved successfully.",
  "data": {
    "user_id": "SooJK37gzjIZfFNH0tlL",
    "device_info": "t5k77PHnuDSrgEzvMJAj",
    "device_id": "3ioui23u29sjkdnsjkndsk",
    "authorization_status": "NOT_INIT",
    "access_token": "skljalkjdlksjlaskjskld"
  }
}
```

GET /auth/qr-code-auth?device_id=InvalidUserData

Status: 404 Not Found
Response
```
{
  "message": "No Authentication found."
}
```
GET /auth/qr-code-auth?user_id=GTB4UUtlKwGemRN2lwBp11&mobileId=3ioui23u29sjkdnsjkndsk

Status: 400 Bad Request
Response
```
{
  "statusCode": 400,
  "error": "Bad Request",
  "message": "invalid query parameters passed"
}
```

GET /?device_id=3ioui23u29sjkdnsjkndsk

Status: 500 Internal Server Error
Response 
```
{
  "message": "The server has encountered an unexpected error. Please contact the administrator for more information."
}
```

In case of any doubt let's have a discussion on a call or you can text the Mobile-App-React-Native channel on discord.
You can also check our[ API contract](https://github.com/Real-Dev-Squad/website-api-contracts/tree/main/qr-code-auth) in case of error responses.

### PR DESCRIPTION
**Pull request by Satyam005**
Closes #1011 and closes 3/3 for #1176 

## Requested Feature
GET API call to retrieve the QR-code-auth document to retrieveEndpoint: GET /auth/qr-code-auth

**_Response_**
image

**_Error 404_**
image

**_Error 400_**
image

### CODE REVIEW

**Comment by isVivek99** [comment]
one test case is failing @Satyam005 , can we check that

File: controllers/auth.js [hunk 1]
```diff
@@ -122,9 +127,32 @@ const updateAuthStatus = async (req, res) => {
   }
 };
 
+const fetchUserDeviceInfo = async (req, res) => {
+  try {
+    const queryParamArray = Object.keys(req.query);
+    if (queryParamArray.length === 1 && queryParamArray[0] === "device_id") {
```
Comment (heyrandhir) [inline controllers/auth.js:133]: please move this logic to validator

File: controllers/auth.js [hunk 2]
```diff
@@ -122,9 +127,32 @@ const updateAuthStatus = async (req, res) => {
...
+      const deviceId = req.query.device_id;
+      const userDeviceInfoData = await QrCodeAuthModel.retrieveUserDeviceInfo(deviceId);
+      if (!userDeviceInfoData.userExists) {
+        return res.boom.notFound("No Authentication found!");
+      }
```
Comment (heyrandhir) [inline controllers/auth.js:138]: error message is very generic

File: models/levels.js [hunk 3]
```diff
@@ -27,7 +27,6 @@ const addLevel = async (levelData) => {
 const deleteLevel = async (id) => {
   try {
     await levelModel.doc(id).delete();
-    return;
```
Comment (heyrandhir) [inline models/levels.js:30]: why are we removing this.

File: controllers/auth.js (hunk 2 above)
Comment (heyrandhir) [inline controllers/auth.js:138]: Also it should be 404

File: test/integration/qrCodeAuth.test.js [hunk 4]
```diff
@@ -160,4 +160,51 @@ describe("QrCodeAuth", function () {
...
+      chai
+        .request(app)
+        .get(`/auth/qr-code-auth?device_id=${userDeviceInfoData.device_id}`)
+        .end((err, res) => {
+          if (err) {
+            return done(err);
+          }
+          expect(res).to.have.status(200);
+          expect(res.body).to.be.a("object");
+          expect(res.body.message).to.equal(`Authentication document retrieved successfully.`);
```
Comment (heyrandhir) [inline test/integration/qrCodeAuth.test.js:186]: can we also assert req.body.data fields

**Review by heyrandhir — CHANGES_REQUESTED** [review]

**Comment by heyrandhir** [comment]
@Satyam005 your tests are failing. Please check.

File: controllers/auth.js (hunk 2 above)
Comment (Satyam005) [reply controllers/auth.js:138]: It's giving 404 only @heyrandhir

File: controllers/auth.js (hunk 1 above)
Comment (Satyam005) [reply controllers/auth.js]: added @heyrandhir

File: controllers/auth.js (hunk 2 above)
Comment (Satyam005) [reply controllers/auth.js]: Replace the message with "User with id <:id> does not exist."

File: models/levels.js (hunk 3 above)
Comment (Satyam005) [reply models/levels.js:30]: it is because of lint fix, it's removing this return statement

File: test/integration/qrCodeAuth.test.js (hunk 4 above)
Comment (Satyam005) [reply test/integration/qrCodeAuth.test.js]: added @heyrandhir

**Review by isVivek99 — APPROVED** [review]

**Review by heyrandhir — CHANGES_REQUESTED** [review]
can we please add tests for validator. rest looks good to me.

**Review by heyrandhir — APPROVED** [review]

**Review by isVivek99 — APPROVED** [review]

**Review by shreya-mishra — APPROVED** [review]

**Pull request merged**